ems_refresh.openshift.store_unused_images setting

cben opened the pull request:
Strawman alternative/baseline for #14628.
Allows to still get_container_images=true in one big request,
but instead of saving metadata (labels etc) on all images openshift gave us,
save it only for images that have been mentioned by pods.

- Images registries are still all saved, even if only mentioned in unused images.

#14628 fetching used images one-by-one may be better eventually, but I think this is a simple knob we'll want anyway, if only for comparison.
The end state here and in #14628 should be same — unused images get disconnected, used images have full metadata.
~~PR based off #14661 test additions.~~

Measuments on real data with 44000 images suggest this is acceptable time-wise (24min vs 11min with get_container_images=false), but eats a lot of RAM (preliminary data: ~2GB while parsing, GC after parse could reclaim ~0.7GB of that, then up to 1GB while saving?).

cben commented:
cc @enoodle @Ladas @agrare @simon3z

Ladas commented:
@cben the parsing might be an issue of the OpenShift client? We are seeing similar thing in Amazon client, where just fetching of the 80k images take like 800MB in memory, so it has some fat objects. (80k images JSON has only like 20MB)

In save inventory, the biggest peaks are in second+ refresh, where we load all 40k ActiveRecord object in memory https://github.com/Ladas/manageiq/blob/7099177c291dadb3588938a2a22032048cc20604/app/models/ems_refresh/save_inventory_helper.rb#L37-L37

agrare commented:
@cben I like this much better than the post-refresh method, we can deal with the memory usage on collection but the save_inventory time when saving container_images was a real killer

agrare commented on app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb:
```diff
@@ -165,10 +167,11 @@ def parse_env_variables(env_variables)
         id = openshift_image[:dockerImageReference] || openshift_image[:metadata][:name]
         ref = "#{ContainerImage::DOCKER_PULLABLE_PREFIX}#{id}"
-        new_result = parse_container_image(id, ref)
+        new_result = parse_container_image(id, ref, :store_new_images => options.store_unused_images)
```
The argument name should match the config key, `:store_new_images` sounds to me like it should do something different than `store_unused_images`

cben replied:
My thinking was that this function is called from two places (pods, images), with potentially different value.  it doesn't have the context to understand which images are unused, only whether it already has it in data_index.
Even when setting is false, while parsing pods we want to add "new" images. IMO it'd be confusing to pass `store_unused_images: true` when the eponymous setting is false.

enoodle commented:
Overall LGTM, I agree with Adam about the argument's name.

About the travis failures it seems like you should also update "spec/models/manageiq/providers/openshift/container_manager/refresh_parser_spec.rb" with the new argument to `get_openshift_images`.

cben commented:
Fixed refresh_parser_spec.  Do you think I should add any unit tests there, or are the full refresher specs enough?

enoodle commented:
@cben Yes, It is also currently using the default from `Settings.ems_refresh.openshift` which is not very clear in the test what values it holds. Will be nice to have one test to check either option so that we don't have to debug through the refresher test if we accidentally break something in the future.

enoodle commented:
cc @zeari

cben commented:
Closing in favor of
https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/11
https://github.com/ManageIQ/manageiq-providers-openshift/pull/9

Added parser tests there.  Replaced `Settings.ems_refresh.openshift` in tests with a struct, see http://talk.manageiq.org/t/settings-defaults-and-tests/2309 for a (long) braindump about that.

durandom commented:
@cben see also https://github.com/ManageIQ/manageiq/pull/14982 and https://github.com/ManageIQ/manageiq/pull/14974 which happens to address the same issue - nice coincidence :)
